### ISSUE TITLE
ngcc modifies node_module making it no longer work in a web worker

### ISSUE THREAD

**Issue by kragoth** [angular/angular#42139]
# Bug Report

### Affected Package

This issue is caused by ngcc (as far as I can tell)

### Is this a regression?

Yes, this previously used to work. I'm not sure exactly what version broke it but, I had this working in angular 10 (or a very early version of angular 11).

### Description
It appears ngcc is modifying files within node_modules/@datorama/akita/lib. The modificaitons add @angular/core as an import in store.d.ts (among a few other lines of code). Once this occurs I am no longer able to build the app if a web worker imports the @datorama/akita package as @angular/core requires the dom library which is not available in a web worker.

## Minimal Reproduction

I have a reproduction here: https://stackblitz.com/edit/angular-ivy-wvs54l
However, stackblitz doesn't seem to support webworkers. So, you need to download the project and then perform:
```
npm i
npm run ng build
```

Essentially all you need to do is create an angular app
* Add the @datorama/akita npm package
* Add a webworker
* Instantiate the webworker in the angular app
* In the webworker import the @datorama/akita package

Once you have done this any attempt to build the app will fail.

## Exception or Error
<pre><code>
Error: ./src/app/web-workers/test.worker.ts (./node_modules/worker-plugin/dist/loader.js?name=0!./src/app/web-workers/test.worker.ts)
Module build failed (from ./node_modules/worker-plugin/dist/loader.js):
Error: node_modules/@angular/core/core.d.ts:10411:62 - error TS2304: Cannot find name 'Element'.

10411 export declare function ɵgetDebugNode__POST_R3__(nativeNode: Element): DebugElement__POST_R3__;
</code></pre>
There are more errors that follow which all are basically due to the dom library not being available in a web worker.

## Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 11.2.13
Node: 14.16.0
OS: win32 x64

Angular: 11.2.13
... animations, cli, common, compiler, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1102.13
@angular-devkit/build-angular   0.1102.13
@angular-devkit/core            7.3.10
@angular-devkit/schematics      7.3.10
@angular/compiler-cli           11.2.14
@schematics/angular             11.2.13
@schematics/update              0.1102.13
rxjs                            6.6.7
typescript                      4.0.7
</code></pre>

**Anything else relevant?**

So, I don't really know how ngcc works but, what I do know is that I end up with 2 files in @datorama/akita/lib
* store.d.ts
* store.d.ts.__ivy_ngcc_bak

If I replace store.ts with store.d.ts.__ivy_ngcc_bak I can now build successfully.

**Comment by kragoth** [comment]
Oh if anyone knows how I could work around this problem I would be very grateful. I'm currently blocked by this problem!

**Comment by petebacondarwin** [comment]
You are right that it is ngcc that is causing this change to the `.d.ts` file that results in `@angular/core` being included in the typings dependencies of the `@datorama/akita` package.

@alan-agius4 suggested that the short-term workaround is to skip checking the typings of libraries for your web-worker build. You do this by updating the `tsconfig.webworker.json` file to be:

```json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "./out-tsc/worker",
    "lib": [
      "es2018",
      "webworker"
    ],
    "types": [],
    "skipLibCheck": true
  },
  "include": [
    "src/**/*.worker.ts"
  ]
}
```

Note the new `"skipLibCheck"` option.

While this is not ideal it only prevents TypeScript from checking the typings in libraries, and only for the web-worker build.

**Comment by petebacondarwin** [comment]
The longer term problem is that, while it is ngcc that is adding the typing dependency, there is still an implicit dependency between Aikita and Angular because the `Store` is marked as an Angular directive. So even moving on to the new partial compiled Ivy library distribution will not solve the problem.

**Comment by petebacondarwin** [comment]
Actually - looking at this a bit more. I don't quite understand why ngcc things that the Akita `Store` class is a directive. This looks wrong to me...

**Comment by petebacondarwin** [comment]
Ahah! It is because the `Store` class has this method:

```
    ngOnDestroy() {
        this.destroy();
    }
```

It seems that we search for methods like this. But perhaps we could be more sophisticated and ignore classes that are not also decorated with Angular decorators...

**Comment by petebacondarwin** [comment]
Here is the code where this method is detected and causes ngcc to consider this class an "abstract" directive.
https://github.com/angular/angular/blob/a86ca4fe049e68710b612a8e3323c256115a805e/packages/compiler-cli/src/ngtsc/annotations/src/directive.ts#L190-L198

I think we should (at the least) provide an option in `ngcc.config.js` to turn off this feature of compilation for certain entry-points. But given that it is now two major versions since this became unsupported, perhaps we could turn this off altogether? Or at least turn it off for ngcc...

**Comment by alxhub** [comment]
> Ahah! It is because the `Store` class has this method:
> 
> ```
>     ngOnDestroy() {
>         this.destroy();
>     }
> ```
> 
> It seems that we search for methods like this. But perhaps we could be more sophisticated and ignore classes that are not also decorated with Angular decorators...

Unfortunately that's the point of this feature - we need to find _undecorated_ classes with such lifecycle methods.

`Store` here is not really an abstract directive, but a service. Unfortunately there's no good way to distinguish between the two, so ngcc just makes an assumption that _most_ such classes are abstract directives. The "correct" change is to decorate `Store` with `@Injectable()`.

**Comment by petebacondarwin** [comment]
Given that akita doesn't actually contain any Angular stuff, then I think we could get round this with an `ngcc.config.js` setup that ignores this package...

**Comment by kragoth** [comment]
@petebacondarwin is the ngcc.config.js something that I can do now or is it a feature that needs to be added to ngcc first?
In the mean time I'll give the "skipLibCheck": true option a go. 
Thank you for the help!

**Comment by kragoth** [comment]
Good news on this actually.
The author of Akita has said that this is actually legacy code and he will remove it in the next major release of Akita.
https://github.com/datorama/akita/discussions/678

The "skipLibCheck": true workaround is working for me currently so thanks for that :)

**Comment by petebacondarwin** [comment]
Great! I think we can close this issue, since this is working by design in Angular, there is a workaround, the library will fix it, and it is very rare.

For your information the ngcc config approach would be to create the following file in your own project (next to `package.json`):

**`ngcc.config.js`**

```ts
module.exports = {
  packages: {
    '@datorama/aikita@0.1.1': {
      entryPoints: {
        '.': {ignore: true},
      },
    },
};
```

(Disclaimer, I have not tested this!)
